### ISSUE TITLE
VMU/Maple support for DreamConn+ controller

### ISSUE THREAD

**Issue by chrisvcpp** [flyinghead/flycast#1305]
Hi there!
My name is Chris Daioglou, and I'm the creator of the DreamConn+ wireless controller, and the later VM2 for the Dreamcast
It's a while now that I was thinking of suggesting fellow developers for an addition to the Flycast emulator.

As you may know, the DreamConn+ is the only wireless controller, that supports actual VMU & Rumble Pak, giving the same experience as the original wired controller. It's a while now, that I've added support for using the controller with Windows PC, making it compatible with Steam, games and emulators.

The only feature that lacks from any current emulation system, is the transmitting of Maple data regarding the VMU, or Rumble Pak.
So, I was thinking that we could collaborate, in order for Flycast (Windows version at least) to transmit/communicate with an intermediate channel (i.e. local IP port) that the DreamConn+ could use to gather data and use with the actual VMU plugged into the controller.
This should give the ultimate experience with using the emulator along with the DreamConn+, giving an experience as close to the original console as possible.

**Comment by flyinghead** [comment]
Hi Chris,
Sorry for not getting back to you earlier (!!)
How do you see an integration of DreamConn+ with Flycast so that maple data can be exchanged? I see that it can be used with a PC, and I guess the controller part works with flycast through SDL.
If using a local IP socket for maple data, would that work with more than one controller? Do you have an idea for the data format/protocol?

**Comment by chrisvcpp** [comment]
Hi,
Thank you for getting back to me!
I actually already did dome progress with it. Please check a video preview here: https://x.com/CDaioglou/status/1852811345684533310
I used the later 2.4 version from the Git and made a custom built - only for Windows x64 since the DreamConnS controller is compatible only with Windows.
At the moment I used something similar to named pipeline(instead of IP socket), and using only the LCD Maple messages for all Slot[1] ports : A1, B1, C1, D1
Didn't commit the changes to the project, as I know that Flycast aims into being cross-platform..
But, it would be great if you'd be willing to officially add support for the DreamConnS controller.
I can imagine that using local IP sockets (one for each port, so 4 in total) would be awesome!
The data format/protocol would be the Maple commands as they're! No need to make any special formatting.
Just forwarding the Maple commands through the corresponding IP port would be more than enough to me.
This way, we can add support for even more functionalities than just LCD data.For example, I could use all of the Maple commands for actually using all of the VMU features : LCD, save/load, buzzer, clock/date, etc.
Let me know if this's something feasible and that you'd like to do..

Many thanks,- Chris


    Στις Τετάρτη 11 Δεκεμβρίου 2024 στις 02:24:39 μ.μ. EET, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 


Hi Chris,
Sorry for not getting back to you earlier (!!)
How do you see an integration of DreamConn+ with Flycast so that maple data can be exchanged? I see that it can be used with a PC, and I guess the controller part works with flycast through SDL.
If using a local IP socket for maple data, would that work with more than one controller? Do you have an idea for the data format/protocol?

**Comment by flyinghead** [comment]
Yeah, I saw your video when you posted it on Twitter. Impressive work!
Is there any way you could share your flycast changes? I could integrate them upstream. I don't have a device to test with so I need at least some working prototype code to boot me up.

**Comment by chrisvcpp** [comment]
Sure thing!
I only added a few lines of code to the maple_devs.cpp -> function u32 dma(u32 cmd) override -> section case MFID_2_LCD :
 ==============================================
case MFID_2_LCD:{     DEBUG_LOG(MAPLE, "VMU %s LCD write", logical_port);     r32();     rptr(lcd_data,192);
     if (!strcmp(logical_port, "A1") || !strcmp(logical_port, "B1") || !strcmp(logical_port, "C1") || !strcmp(logical_port, "D1"))     {         char sfile[30]; sprintf(sfile, "data\\vmu_lcd_%s.dat", logical_port);         FILE* f = std::fopen(sfile, "wb");         if (f != NULL)         {             std::fwrite(lcd_data, 1, 192, f); std::fflush(f); std::fclose(f);         }     }    .
     .
    .
==============================================
As you can see, I'm just writing the 192 bytes of the LCD-related data to a corresponding A1/B1/C1/D1 file - the rest of work is done by the DreamConn utility app..
In our case, working with IP ports will be much easier (and a better solution) : you will just have to forward the Maple messages related to VMU(A1/B1/C1/D1) to the IP port.Then I will handle everything through my application..


    Στις Τετάρτη 11 Δεκεμβρίου 2024 στις 03:30:16 μ.μ. EET, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 


Yeah, I saw your video when you posted it on Twitter. Impressive work!
Is there any way you could share your flycast changes? I could integrate them upstream. I don't have a device to test with so I need at least some working prototype code to boot me up.

**Comment by flyinghead** [comment]
Would it make sense for the DreamConn utility app to be the "network server", to which flycast would connect to?
I suppose your utility must be running as long as you use a DreamConn+ controller. Whereas flycast may be opened and closed several times depending on which game the user is playing.

So the DreamConn utility could open a listening socket (or one socket per controller) on known ports. Flycast would then try to connect to this port at start up, and establish the connection. Then raw maple data would be sent to it.

A few more thoughts:
It would help if flycast could detect that a given controller is a DreamConn+. That way flycast will only connect to the DreamConn utility when a DreamConn+ is detected, and correctly identify which maple port it's connected to. I guess a vendor/product ID is enough, or the controller name as reported by SDL.

As a first step, I can forward maple messages related to LCD screen update and beeper. These messages don't really need any answer from the controller or vmu. However, if we want to use other functions such as storage, flycast will have to wait for a maple response before proceeding.

Flycast will also need to know the controller configuration: which device is connected (vmu, rumble pack, none) on which port. I guess this can be done with the maple protocol on the network port.

**Comment by chrisvcpp** [comment]
Hi,
Yes, I could make the utility app to be the server. Whenever a controller is connected, I can open a specific TCP port (which corresponds to A1/B1/C1/D1).
Regarding your thoughts:
1. At the moment, when a DreamConn controller is connected, the utility app presents a virtual XBox360 controller (XInput) to Windows (up to 4 of them).This virtual controller can also report specific vendor/product ID, so yes, this is quite feasible and easy to do!
2. You're quite right about this! Maple messages that do not require specific responses (other than "ACK") like LCD & BUZZER will be easier to begin with.For other more complex Maple commands (like saving/loading), we can do it later on - I can think of a few "tricks" that can make it easier and also keep Flycast as close to its original/current state as possible.
3. For the controller configuration, this can easily be done by Flycast sending a "Get Condition" (command 9) to the TCP port.The utility will respond with the controller's state indicating which slots are occupied (Slot1 occupied means VMU is present, Slot2 occupied means Rumble is present).

    Στις Πέμπτη 12 Δεκεμβρίου 2024 στις 06:02:29 μ.μ. EET, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 


Would it make sense for the DreamConn utility app to be the "network server", to which flycast would connect to?
I suppose your utility must be running as long as you use a DreamConn+ controller. Whereas flycast may be opened and closed several times depending on which game the user is playing.

So the DreamConn utility could open a listening socket (or one socket per controller) on known ports. Flycast would then try to connect to this port at start up, and establish the connection. Then raw maple data would be sent to it.

A few more thoughts:
It would help if flycast could detect that a given controller is a DreamConn+. That way flycast will only connect to the DreamConn utility when a DreamConn+ is detected, and correctly identify which maple port it's connected to. I guess a vendor/product ID is enough, or the controller name as reported by SDL.

As a first step, I can forward maple messages related to LCD screen update and beeper. These messages don't really need any answer from the controller or vmu. However, if we want to use other functions such as storage, flycast will have to wait for a maple response before proceeding.

Flycast will also need to know the controller configuration: which device is connected (vmu, rumble pack, none) on which port. I guess this can be done with the maple protocol on the network port.

**Comment by flyinghead** [comment]
I committed a first prototype for DreamConn+ support (`dev` branch)
Right now, each dreamconn must be enabled in emu.cfg:
```
[input]
DreamConn0 = yes
DreamConn1 = yes
DreamConn2 = yes
DreamConn3 = yes
```
At start up, flycast will connect to locahost:37393 for port 0, localhost:37394 for port 1 and so on.
it will send a `MDCF_GetCondition`/`MFID_0_Input` maple message: `09 20 00 01 00 00 00 01` and wait for a reply, which is mostly ignored except to determine which expansion devices are connected to the controller.

When a game starts, flycast will forward maple messages for the VMU LCD and rumble pack to the DreamConn utility through the network socket. `MDCF_BlockWrite`/`MFID_2_LCD` for the VMU LCD and `MDCF_BlockWrite` and `MDCF_SetCondition` for the rumble pack. It doesn't expect a response in this case and won't even try to read from the socket (at least for now).

**Comment by chrisvcpp** [comment]
Great! Will start modifying the Utility app and let you know how it goes..
Also, (in case it's of any use to you) I've already modified the Utility to register each virtual XBox360 controller with :Vendor ID: 0x4457Product ID : 0x4443

    Στις Δευτέρα 16 Δεκεμβρίου 2024 στις 07:15:39 μ.μ. EET, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 


I committed a first prototype for DreamConn+ support (dev branch)
Right now, each dreamconn must be enabled in emu.cfg:
[input]
DreamConn0 = yes
DreamConn1 = yes
DreamConn2 = yes
DreamConn3 = yes

At start up, flycast will connect to locahost:37393 for port 0, localhost:37394 for port 1 and so on.
it will send a MDCF_GetCondition/MFID_0_Input maple message: 09 20 00 01 00 00 00 01 and wait for a reply, which is mostly ignored except to determine which expansion devices are connected to the controller.

When a game starts, flycast will forward maple messages for the VMU LCD and rumble pack to the DreamConn utility through the network socket. MDCF_BlockWrite/MFID_2_LCD for the VMU LCD and MDCF_BlockWrite and MDCF_SetCondition for the rumble pack. It doesn't expect a response in this case and won't even try to read from the socket (at least for now).

**Comment by chrisvcpp** [comment]
I've made the modifications to the Utility and now ready to test :
1.  Do you have a build somewhere that I can download, or do I have to build it myself?
2. One minor change I would suggest is this : instead of transmitting the Maple messages in raw bytes, it would be better if you could transmit each message as a string with line feed at the end (\r\n). This will make it easier for me to distinguish between messages and buffer them in a FIFO pipeline. The bytes can be represented in HEX with a space separator.

**Comment by flyinghead** [comment]
You can download the lastest `dev` build here: https://flyinghead.github.io/flycast-builds/#dev

So you basically want the message as an ASCII dump like in this C string `"09 20 00 01 00 00 00 01\r\n"`. I assume you will use the same format for responses.

**Comment by chrisvcpp** [comment]
That's right, the messages can be in the format : "09 20 00 01 00 00 00 01\r\n"
For response, I can use any format that suits you - either the same string format, or raw bytes..
Once you have the updated build, please let me know so I can download it and start testing..

**Comment by flyinghead** [comment]
New build available:
Sending and receiving data as text.
Should be able to detect DreamConn+ gamepad automatically.

**Comment by chrisvcpp** [comment]
Great!
So, no need to enable DreamConn through .cfg any more? (Like, used the Vendor/Product ID?)

**Comment by flyinghead** [comment]
yes, I removed the emu.cfg config entries so I hope VID/PID detection works...

**Comment by chrisvcpp** [comment]
Awesome! Will make the tests and update you further.

**Comment by chrisvcpp** [comment]
Just started with the tests. So far communication with Flycast does not seem to be working.
Once a DCS controller is connected, the Utility does the following actions:

1. Creates a virtual XBox360 controller with VID: 0x4457, PID: 0x4443 (attached screenshot of device properties)
![DCS_Dev](https://github.com/user-attachments/assets/72ee0347-8f23-4050-8029-944c686a0d51)
2. Opens a listening socket according to the selected Port (A: 37393, B: 37394, ..)

After the above actions, I'm running Flycast (dev build b9fdd50) but server status shows that no client is connected..

**Comment by chrisvcpp** [comment]
Discard my previous message. Connection works just fine!
I was just expecting to read the Maple message in the same order as it runs in the actual bus (BIG ENDIAN)
Seems that you're sending it in LITTLE ENDIAN order, so it's OK..

**Comment by flyinghead** [comment]
This is the order in which data is usually sent to the dreamcast maple interface.  But it can be changed with a register and some diagnostic disk actually sends data in big endian mode.
Anyway, glad to hear that it's working so far.

**Comment by chrisvcpp** [comment]
In low-level, the bytes are actually inverted. Working for so many years directly with the Maple Bus, I'm used into reading it in BENDIAN 😄 No worries, I will handle the order through the Utility.
For now, everything seems to be working as expected! 👍 
Will upload some footage soon..

**Comment by chrisvcpp** [comment]
OK, so far everything works great! One thing I noticed though:
It seems like Flycast sends the same (current) LCD message multiple times.
I guess there is something like a timer which constantly sends the last message?
I would suggest that it transmits to the TCP port only when a new message arrives.
This way we can avoid possible stuttering on the Flycast side and also latency (buffering) to the controller itself..

**Comment by flyinghead** [comment]
Flycast only transmits messages when received from the game. There's no timer to resend anything.
The only exception is when loading a state, where the screen is refreshed right after loading.

**Comment by chrisvcpp** [comment]
Strange.. After the 1st LCD message is transmitted, the server is constantly receiving - even when no LCD change has occurred.. 🤔

**Comment by chrisvcpp** [comment]
Yep. I also tested with a game using static LCD (i.e. Half Life).
After the LCD message arrives, the server keeps receiving the same message again & again.
This does not happen with the initial "Get Condition" message though..
So, I made this thought :
Is there a chance that the underlying socket in Flycast keeps sending the same message because you don't read after transmitting?
For example, if you make it read a generic "ACK\r\n" message after transmission? This could make this bug disappear..?

**Comment by flyinghead** [comment]
No, it can't come from the tcp/ip socket itself, and reading from and writing to a socket are disconnected.
I tested again and I only see messages when the VMU LCD is modified by the game. I used 240pSuite for this test.
(I'm using a test tcp/ip server that dumps everything it receives).

Are you sure it's not coming from the network library or framework if you're using any?
You can recompile flycast and add some logging for each message sent over the network.

**Comment by flyinghead** [comment]
Tested Half-Life and I only see one LCD update when the game boots.

**Comment by chrisvcpp** [comment]
Yes, actually you're right! Tested with Half-Life again, and actually sends only one message.
Then launched RE:CV, and from the moment the 1st LCD came in, it started receiving the same message all over again...
If you launch RE:CV, do you see everything to be in normal?
If so, then I suppose it's some kind of bug with C# framework on my end..

**Comment by flyinghead** [comment]
RE:CV is updating the LCD screen every other frame (15 times per second) so what you're seeing is normal.

**Comment by chrisvcpp** [comment]
Man... I knew it did this when saving, but didn't occur to me doing it constantly! 😄 
So, false alarm here! The network works as expected in both ends👍
I just have to make some "tricks" to the Utility as the controller over BT cannot of course handle such timings..

Oh, another thing : if you like you can also add MDCF_SetCondition for the CLOCK.
This will enable the buzzer support and does not require a response too.

**Comment by chrisvcpp** [comment]
OK, just an update : Everything works perfectly!
Both VMU/LCD and Rumble Pak messages are coming properly from Flycast and forwarded to the controller accordingly.
Great job! 👍

**Comment by flyinghead** [comment]
Pushed a new build on the `dev` branch with support for buzzer messages.
I used a new network library so there might some regressions. Let me know if anything is broken.

**Comment by chrisvcpp** [comment]
Awesome! Will test now and let you know, thanks! 👍

**Comment by chrisvcpp** [comment]
Just tested - everything works perfectly! 👌 

BTW, another thing I'd like to ask : I had this thought of adding a hotkey to the DreamConn which would make Flycast bring-up the main menu. So, I tried to map L+R+START and make the Utility send the "Tab" key to Flycast.
Unfortunately, Flycast seems "immune" to any existing methods of virtually sending key strokes.
So, I thought if it's possible to do it from within the emulator itself : if it detects that L+R+START are pressed, then act like the Tab key was pressed. I think it will be a nice feature to the user, being able to access the menu from the controller without the need of the keyboard..

**Comment by flyinghead** [comment]
Flycast uses SDL to handle all input on Windows, so it *should* react to any keyboard event. SDL might be getting in the way though but I don't know how and why.
Does the standard controller mapping work with the DreamConn+? L, R and Start are correctly mapped by default?

**Comment by chrisvcpp** [comment]
Yes, the mappings on the controller are OK. When I detect L+R+START, I'm sending a virtual {Tab} press to the active window.
I tested with other applications, (i.e. Notepad) and works fine - it just can't be recognized by Flycast.
As you said, I also suspect that something's has to do with SDL preventing the standard virtual key strokes..
Also, since the Utility is written in C#, I can't use SDL (through which it could possibly work).
So, seems the Utility can't do the "trick", and I thought that another solution could be doing it from within the emulator..

**Comment by chrisvcpp** [comment]
Oh, something I just noticed with this version : seems that with the new network library, some messages are randomly "broken".
The server receives lines of messages that seem to be smaller than should, and then receives a new line that seems to be part of the previous message, and so on.. With the previous build this was not happening and all messages were coming in one line, as expected.
For now I'm just ignoring the wrong lines, but the result is less smooth that with the previous build..

**Comment by flyinghead** [comment]
When you say that some messages are “broken”, do you mean that there’s a CR/LF in the middle of the line? Or just that your read() returns an incomplete line, with the rest returned in a subsequent read()?
In the former case, it’s a Flycast bug. In the latter, you should continue reading until you get a new line sequence.

**Comment by chrisvcpp** [comment]
The server always receives until a new line feed is received, otherwise it returns nothing.
So, I can assume that some messages contain a CR/LF somewhere in the middle, causing the server to return these "false" messages.
Since, I've changed nothing on the server side, I guess that the new client network library you mentioned works in a different way, than the previous one (which worked without this issue)..?

**Comment by flyinghead** [comment]
I see that there is a lot of fragmentation and data is sent with a lot of very small packets (some with only a few bytes).

However the stream is correct and there's no spurious `\r\n` being sent in the middle of a line.

I will buffer on my side and send the complete message in one write to improve this but again, there's no guarantee that a given data line will be sent in a single packet (and thus can be read in one read()). Relying on this will break at some point or another.

If you see invalid data being sent, some log or dump of the received data would be helpful.

**Comment by chrisvcpp** [comment]
I see. If there's a lot of fragmentation, then indeed this could cause issues..
Fragmentation is OK when working with raw bytes, but in our case we work with formatted string representation.
That would be great if you could buffer and send each message in one write.
I'll also see what I can do on the server side and do some buffering as well (if possible), and also send you a log (in case it won't 100% work)..

**Comment by flyinghead** [comment]
New build underway. Should fix the fragmentation issue but again, I strongly recommend not to rely on reading a full maple message in a single read. The maximum size of a maple message is 1024 bytes. When translated into ASCII it becomes 3073 bytes, which is larger than the maximum IP packet size on most networks and in many OSes.

**Comment by chrisvcpp** [comment]
Great!
Well, I did some more research and I attach a log of the lines that server receives (tested with RE:CV - LCD data only).
There, you can see the "proper" lines which start with "0C" and the "false" lines which are of random data and size.
As it seems, those "false" lines are not part of a previous (or next) line.
That is because in most cases the previous or next line is a properly received maple message.
So, I can only assume that those lines are part of individual messages that were not properly received for some reason..
Anyway, hope this helps you in any way. Please note though that the initial build was working fine.
I'll continue to investigate this on my end, until the new build is ready 👍  Thanks!
[Maple.log](https://github.com/user-attachments/files/18266054/Maple.log)

**Comment by flyinghead** [comment]
One more build this time with the LT+RT+Start combo handling.

**Comment by chrisvcpp** [comment]
You're the man! Tested the latest build and works like a charm! Not a single message/line is dropped and the addition of L+R+Start combo is great! Excellent experience now! 👍 

Don't know whether or when this should be transferred to the official build of Flycast - this is in your discretion 😃 
BTW, except for writing this awesome emulator, do you also play with actual Dreamcast as well?
Please, p.m. me through my website...

**Comment by flyinghead** [comment]
These changes are now available on the master branch.

**Comment by chrisvcpp** [comment]
Great! Thank you very much!
Also, tested with more controllers connected at the same time, and everything works perfectly! 👌

**Comment by chrisvcpp** [comment]
Thank you!
One thing I noticed just yesterday : when using game states (saving/loading) seems like Flycast stops sending Maple messages to the DreamConn.
Closing the game and starting again, seems that brings it back to normal (until a state is used again)..

    Στις Τετάρτη 15 Οκτωβρίου 2025 στις 07:41:01 μ.μ. EEST, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 


Closed #1305 as completed.

**Comment by flyinghead** [comment]
All right, I'll have a look. There's been quite some changes in this area lately.

**Comment by chrisvcpp** [comment]
Yeah, I know..
Thanks! Let me know if you need any assistance on my end.

    Στις Τετάρτη 15 Οκτωβρίου 2025 στις 08:07:33 μ.μ. EEST, ο χρήστης flyinghead ***@***.***> έγραψε:  
 
 flyinghead left a comment (flyinghead/flycast#1305)
All right, I'll have a look. There's been quite some changes in this area lately.

**Comment by chrisvcpp** [comment]
Oh, btw, while you are at it, there's a minor refinement that could be done.
It has to do with the **L+R+START** shortcut for bringing the Flycast menu up :

At the moment, if L+R+START are pressed, the menu will show but it will continuously toggle on/off while the buttons are pressed.
So, I thought of this approach:

1. Pressing L+R+START shows the Flycast menu
2. Wait for the user to release L+R+START (or just one of them)
3. Continue with the menu
4. If user presses L+R+START again, hide the Flycast menu
5. Wait for the user to release L+R+START (or just one of them)
6. Continue with emulation

**Comment by flyinghead** [comment]
I can't seem to reproduce either problem:
I tested with Crazy Taxi and even after saving or loading a state, the VMU display is still animating. I did several tries but the VMU display is still being updated after a load or save.
I also tested the pause menu shortcut and keeping L+R+Start pressed has no effect for me. I must release them first for the shortcut to work again. Same behavior for pausing and unpausing.

Tested on Windows 10 with latest master build (v2.5-148-g3d594054e) and DreamConnS Utiility 2.4.0.0.

**Comment by chrisvcpp** [comment]
For the save/load state, what occurred to me was when I was testing with Prince of Persia 3D :
When saving/loading through the game's menu, the VMU produces a beep sound - which did not after saving/loading a state.

OK, I will test again with the latest build (I was testing with v2.5-37 from Sep-12th) and let you know.
Thank you!

**Comment by chrisvcpp** [comment]
Yep, I was testing with the older build.. 
Now tested with the latest build and works flawlessly! 👍

**Comment by flyinghead** [comment]
Great news!
Closing the issue again. Feel free to reopen as needed.

**Comment by chrisvcpp** [comment]
Hi there!

Just wanted to let you know that I tested the later CI Builds (f65907a / 27a913f) and seems like DreamConn S compatibility is broken.
The DreamLink status shows as connected, but nothing comes to the VMU.

Tested with older versions (8edd8fc) and everything works fine.

**Comment by RikkiGibson** [comment]
It's possible that #2094 broke things, as it made a change to the "protocol", to send an `MDC_DeviceRequest` message to get information about the specific device in a given slot. Do you happen to know whether the DreamConn software works with that?

If you can verify that 4bb3246 fails, but its direct predecessor 2b1eb55 works, that would give good evidence that #2094 was the problem.

**Comment by flyinghead** [comment]
Reverting this change does help. The other problem is `Use Physical VMU Memory` is enabled by default, and DreamLinkVmu interferes with VMU writes and deadlocks the app.

**Comment by chrisvcpp** [comment]
I can also confirm this. Also, noticed that with the later (broken) builds, at startup Flycast shows both Slot1 + Slot2 as empty.
With previous (working) versions, Flycast polls the DreamConn and presents the Slots based on the Maple response coming from the controller. So, something's interfering and looks like overriding or entirely ignoring the controller..

**Comment by flyinghead** [comment]
I found another regression with DreamConn(Imp): it calls `disconnect()` in its destructor, which calls `shared_from_this()`. This method can't be called in the constructor and destructor and causes a `std::bad_weak_ptr` exception to be thrown.

Apparently a consequence of 909ea496ac96c7dc6ba45a8a3a93ca0d7066579a.

**Comment by chrisvcpp** [comment]
I have also reviewed [#2094](https://github.com/flyinghead/flycast/pull/2094) and I can confirm that the DreamConn does not support the use of VMU in Slot2 - it accepts only Rumble Pak.

**Comment by RikkiGibson** [comment]
> DreamConn does not support the use of VMU in Slot2 - it accepts only Rumble Pak.

Is that a hard limitation of the device? Or something you would consider supporting in future depending on e.g. user feedback?

I opened #2172 to try and address the compat issue. Please give the [CI build](https://github.com/flyinghead/flycast/actions/runs/20441345199/artifacts/4946658205) from the PR a try.

If the PR build does not address the problem, please let me know if any relevant info/warn logs are written out when using the DreamConn, and we can go from there.

**Comment by chrisvcpp** [comment]
No, it's not a hard limitation; just a choice of design. You see, when used with an actual Dreamcast, the DreamConn provides the user with a virtual VMU for Slot2. So, there's no need for a real VMU there. I made it this way, so the user can use a Rumble on Slot2 when he wants to.

BTW, I tried the CI build from the PR, but unfortunately nothing changed. The "Physical VMU" is always ON on startup and the comms between Flycast and the DreamConn are still broken.

**Comment by RikkiGibson** [comment]
That's unfortunate, could you please share the logs?

> The "Physical VMU" is always ON on startup

It sounds like there is a bug related to the physical VMU memory support, but, I do not think that the default value for this setting has changed since the last known good build in October.

It's possible that just sending the DeviceRequest message is causing the connection to fail. If there's no interest in DreamConn+Flycast supporting a configuration aside from "VMU in 1, jump in 2", then we could adjust things to never send the DeviceRequest message to the DreamConn controller client.

**Comment by chrisvcpp** [comment]
Sure, here it is.

[flycast.log](https://github.com/user-attachments/files/24304318/flycast.log)

The initial design was Flycast to send messages that only needed an "ACK" response (LCD,Buzzer,Rumble) which in this case the code actually never bothered with responses from the DreamConn Utility - except for the initial StatusRequest (0x09).

I think here's where the issue may be : sending a DeviceRequest(0x01) actually expects a response (which never comes) and due to that the whole comms collapse??

I'd suggest to initially test and stick with the original design ("VMU in 1, Jump in 2") - see if that fixes things..

**Comment by RikkiGibson** [comment]
Thanks!

> I think here's where the issue may be : sending a DeviceRequest(0x01) actually expects a response (which never comes) and due to that the whole comms collapse??

Yes, this sounds likely. I think the logs support that. The thread just waits for a reply to the message. This could present as the emulator locking up, or the connection timing out, etc.

I have pushed a change, with a new [CI build](https://github.com/flyinghead/flycast/actions/runs/20452840534/artifacts/4950304817), to ensure that we don't send the new message type to a DreamConn controller.

I think there's a need to decide how adding new message types over time should work for DreamLink devices, and agree on expectations. There's several potential solutions for that with different tradeoffs.

**Comment by chrisvcpp** [comment]
Yep! That was it. Now works as expected.

For future implementations, maybe we should go with separate TCP ports?

For example, leave the current DreamLink/DreamConn reserved TCP ports as they initially were.
And for new devices/implementations go with new ports? This will maybe make it more flexible and app-specific?

**Comment by flyinghead** [comment]
I still have a deadlock when `Use Physical VMU Memory` is enabled (the default). Since the DreamConn doesn't support VMU writes, this setting should be ignored when dealing with DreamConn.

I also get a segmentation violation on exit when DreamConnImp calls `disconnect()`, which calls `maple_ReconnectDevices()` after most app memory has been freed. I don't think this is needed for DreamConn since maple devices don't change if connectivity is lost. Only the VMU screen won't be updated and vibrations will stop.

Also I'm testing on the `master` branch so please target this branch for your PR. We'll merge back the changes to `dev` but master needs to be fixed first. Thanks

**Comment by Tails86** [comment]
I have a couple of thoughts:
- If the device doesn't support memory, I suggest changing `getFunctionCode()` to not return memory and change `DreamLinkVmu` generation to check for either screen or memory. Then construct `DreamLinkVmu` with or without hardware memory access. I can throw this together if that would help.
- Perhaps the "network expansion device" should be its own, specific implementation rather than piggyback on dreamconn? Or it could extend from and override some things for now?

**Comment by chrisvcpp** [comment]
I agree, and this's what I suggested earlier. The DreamConn implementation maybe should remain untouched.
Instead, new implementations can be applied on other TCP ports - this will make it flexible and allow for future apps to easily adopt to any specific functionality by just communicating through the appropriate TCP port?

**Comment by Tails86** [comment]
Ah! I didn't read between the lines. Yeah, that sounds reasonable to me

**Comment by RikkiGibson** [comment]
Separate TCP ports could work. Just bump the port number by 4 when you want to make a breaking change, essentially.

Then when Flycast wants to connect a DreamLink for a port, it would start by trying to connect to the newest one it knows about. It would fall back to older ones if needed, and use the set of behaviors corresponding to the port. This way a device server can just open the port that it wishes to support, and, if desired adopt a newer version of the protocol later on, without needing to make a simultaneous update to Flycast.

---

**To get things unblocked** I will PR a change to master using `isForPhysicalController()` and similar checks. Later can try to extract out the "version 2" behaviors to a protocol on a new port, etc. At that time I might try to doc what the protocol behaviors actually are.

> I still have a deadlock when `Use Physical VMU Memory` is enabled (the default). Since the DreamConn doesn't support VMU writes, this setting should be ignored when dealing with DreamConn.

I see, this might mean that the VMU write requests are being sent over the socket and no replies are ever received for them. I think that would cause the main emulation thread to just block. I'll see if there's an easy way to make the adjustment you are suggesting.

**Comment by RikkiGibson** [comment]
I think it would be good for DreamLink server(s) in general to reply with an error response of some kind, when they receive an unknown/unsupported message. That way if we accidentally started sending a server a bad message again in the future, there would be a more clear symptom.

**Comment by RikkiGibson** [comment]
> I also get a segmentation violation on exit when DreamConnImp calls `disconnect()`, which calls `maple_ReconnectDevices()` after most app memory has been freed. I don't think this is needed for DreamConn since maple devices don't change if connectivity is lost.

It's probably true the `maple_ReconnectDevices()` call is not needed for the controller scenario, but, it feels like we could also get there in a `!isForPhysicalController()` scenario. e.g. emulator is shutting down, as part of that it disconnects the dreamlinks, and the order of the shutdown happens to be that it freed the maple-related memory already. Would it be reasonable to somehow detect that the emulator is shutting down and therefore we shouldn't call `maple_ReconnectDevices()`?

**Comment by Tails86** [comment]
I created this draft PR regarding my suggestion of disallowing storage at the dreamconn implementation: https://github.com/flyinghead/flycast/pull/2175/commits/d4710ce8bc1e6e5f6977b6511c2134c3056d0ea2

This should ignore whatever that `Use Physical VMU Memory` checkbox is because it's not supported at the hardware level. I have no way to test this though.

**Comment by RikkiGibson** [comment]
If I am understanding right, I think "forcing" `Use Physical VMU Memory: false` for DreamConn controllers is the right thing here. Essentially user likely wants to use Flycast's built-in VMU memory, rather than having the device report that it doesn't have storage.

**Comment by flyinghead** [comment]
@RikkiGibson the best way to check for this is to register for `Event::Start`/`Event::Terminate`.
`maple_ReconnectDevices` should only be called between these 2 events.

**Comment by Tails86** [comment]
> If I am understanding right, I think "forcing" `Use Physical VMU Memory: false` for DreamConn controllers is the right thing here. Essentially user likely wants to use Flycast's built-in VMU memory, rather than having the device report that it doesn't have storage.

My general desire is to keep the `DreamLink` interface generic rather than put in workarounds for a specific device. I feel like we're on the same page when it comes to having something forcefully change a setting on the UI - it can lead to confusion as well. My logical conclusion was that we should treat that checkbox as "use physical memory when available". It's not like physical memory can be used when a virtual VMU is selected. 

These are just my thoughts on it. I'm sure a compromise can be reached on this one way or another.

**Comment by RikkiGibson** [comment]
That's understandable, thanks, and happy to discuss more, and arrive at a more holistic view of how things should work.

I will try to get back on this and maybe adjust the handling of `Event::Terminate` after Christmas. If you find the PR solves the immediate problem as-is, feel free to take it, otherwise I will return to it when available.

**Comment by Tails86** [comment]
I'd personally entertain the idea of completely hiding that checkbox unless something is connected that can possibly handle physical memory. Would something like that go against any established conventions, @flyinghead ?

**Comment by chrisvcpp** [comment]
Hey guys... sorry haven't responded yet; was away for a while - you know, Holidays time :-)

Personally, I'm totally fine with whatever you decide to go with and how the DreamLink should work.

Although, I have a strong feeling that using multiple TCP ports for the DreamLink will serve very nice.
Since a server can have multiple ports at the same time and this way combine functionality according to what a port is bound to serve.

Also, agree with Rikki that maybe the server should respond on an unsupported message; or even better: a response timeout should be applied and treated as "error" or "message not supported". And just continue..

Anyway, I will test tomorrow the latest PR and let you know how it goes with the DreamConn S.

**Comment by flyinghead** [comment]
@Tails86 In general, I favor disabling options that can't be changed or used at the moment, rather than completely hiding them.
The latter makes the UI change dynamically depending on the state, which can be confusing for users.
But even without hiding or disabling it, ignoring it is fine if it doesn't apply to the current state.

**Comment by RikkiGibson** [comment]
> a response timeout should be applied and treated as "error" or "message not supported". And just continue..

I would prefer to avoid conditioning behavior based on a timeout when possible. The code paths which process these messages don't have a lot of spare time, so to speak, to wait for a timeout in order to decide what to do. Such approaches may appear to the user as delays/hitches in the UI, as we may not be able to do the task the game needs in order to proceed, until we get an answer. Deciding on the appropriate timeout in order to fall back in behavior, is not a simple problem, in my opinion, especially as the time needed for the response to arrive could depend on system load and hardware capabilities.

Also, if I wanted to attach a debugger to one or the other process, I would prefer that hitting a breakpoint and inspecting state for a while, then resuming, to *not* be treated by the other peer as: they must not support the operation, I better move on.

**Comment by chrisvcpp** [comment]
No problem. At this point, I could also modify the DreamConn Utility app. How would it be if I send a response **"UNSUPPORTED\r\n"**
Will this be of any help?

Also, @Tails86 , can you please send me an executable with the [d4710ce](https://github.com/flyinghead/flycast/commit/d4710ce8bc1e6e5f6977b6511c2134c3056d0ea2) to test with?

**Comment by RikkiGibson** [comment]
Definitely. I think sending back something like that would be reasonable.

If possible, it would be great if the message were something like: “ERROR: Unsupported function code 0x42.\r\n”.

There might be an interval between sending an unsupported message and actually looking for a reply about some (possibly entirely unrelated) message. So getting that back will give us the necessary info to work out what we did wrong. We could even log such messages and make it easy to tell what’s wrong if this situation arises again in the future.

**Comment by chrisvcpp** [comment]
Sure. Will get on with it.

The message will specifically be : **“ERROR: Unsupported function code 0x42.\r\n”**

Worth noting though : in case of the DreamConn, an unsupported message (i.e. DeviceRequest) shouldn't be considered as "empty Slot".
As per the original design, Flycast should give priority to the StatusRequest(0x09) response : occupied Slots should default to the scheme "VMU on Slot1 / Rumble on Slot2".

**Comment by chrisvcpp** [comment]
Furthermore, in our initial implementation with @flyinghead , Flycast can tell whether a DreamConn is connected by detecting the presense of an XBox360 controller with the following : **Vendor ID: 0x4457 - Product ID : 0x4443**

So, this may also be helpful to you - in order to tell whether the TCP port is listening to a DreamConn or not..

**Comment by Tails86** [comment]
> No problem. At this point, I could also modify the DreamConn Utility app. How would it be if I send a response **"UNSUPPORTED\r\n"** Will this be of any help?
> 
> Also, [@Tails86](https://github.com/Tails86) , can you please send me an executable with the [d4710ce](https://github.com/flyinghead/flycast/commit/d4710ce8bc1e6e5f6977b6511c2134c3056d0ea2) to test with?

Yeah, the builds are available if you look through the Actions tab in flycast. Here is the direct link to that build (look under Artifacts): https://github.com/flyinghead/flycast/actions/runs/20470277151

**Comment by chrisvcpp** [comment]
Unfortunately, that build didn't work

**Comment by flyinghead** [comment]
@chrisvcpp I didn't test the build but Rikki's patch works for me so I merged his PR.
The next `master` build will have this fix so let us know if you find any issue.

**Comment by chrisvcpp** [comment]
Great, thanks!
